fhir-smart/src/main/java/com/ibm/fhir/smart/AuthzPolicyEnforcementPersistenceInterceptor.java

```diff
@@ -695,51 +713,50 @@ private void enforce(String resourceType, String resourceId, Resource resource,
      */
-    private boolean isAllowed(String resourceType, String resourceId, Resource resource, Set<String> contextIds, Permission requiredPermission, List<Scope> approvedScopes)
+    private boolean isAllowed(String resourceType, String resourceId, Resource resource, Set<String> contextIds, Permission requiredPermission, Map<ContextType, List<Scope>> grantedScopes)
             throws FHIRPersistenceInterceptorException {
         Objects.requireNonNull(resourceType, "resourceType");
         Objects.requireNonNull(resourceId, "resourceId");
         Objects.requireNonNull(contextIds, "contextIds");
 
-        Map<ContextType, List<Scope>> approvedScopeMap = approvedScopes.stream()
-                // First filter the list to only scopes which grant the required permissions on the passed resourceType
-                .filter(s -> s.getResourceType() == ResourceType.Value.RESOURCE ||
-                        s.getResourceType().value().equals(resourceType))
-                .filter(s -> hasPermission(s.getPermission(), requiredPermission))
-                // Then group the scopes by their context type
-                .collect(Collectors.groupingBy(s -> s.getContextType()));
-
-        if (approvedScopeMap.containsKey(ContextType.SYSTEM)) {
-            if (log.isLoggable(Level.FINE)) {
-                log.fine(requiredPermission.value() + " permission for '" + resourceType + "/" + resourceId +
-                    "' is granted via scope " + approvedScopeMap.get(ContextType.SYSTEM));
+        for (ContextType context : List.of(ContextType.SYSTEM, ContextType.USER)) {
+            // For `system` and `user` scopes, we grant access to all resources of the requested type.
+            // Implementers that use these scopes are encouraged to layer on their own permissions model beyond this.
+            if (!grantedScopes.containsKey(context)) {
+                continue;
             }
-            return true;
-        }
 
-        if (approvedScopeMap.containsKey(ContextType.USER)) {
-            // For `user` scopes, we grant access to all resources of the requested type.
-            // Implementers that use these scopes are encouraged to layer on their own permissions model beyond this.
-            if (log.isLoggable(Level.FINE)) {
-                log.fine(requiredPermission.value() + " permission for '" + resourceType + "/" + resourceId +
-                    "' is granted via scope " + approvedScopeMap.get(ContextType.USER));
+            Optional<Scope> approvingScope = grantedScopes.get(context).stream()
+                    .filter(s -> s.getResourceType() == ResourceType.Value.RESOURCE || s.getResourceType().value().equals(resourceType))
+                    .filter(s -> hasPermission(s.getPermission(), requiredPermission))
+                    .findAny();
+            if (approvingScope.isPresent()) {
+                if (log.isLoggable(Level.FINE)) {
```

Comment:
I noticed that in other places we're not logging which scope resulted in the request being approved. Perhaps approvals should be logged more broadly with FINER?

Reply:
We don't have a good "single spot" to add this logging, but I could add it to the "isApprovedByScopes" method which is used from a number of thee `beforeX` methods.  I think log level FINE is ok for these and would be consistent with the other spots.

Reply:
added